accepted/future-releases/2509-private-named-parameters/feature-specification.md

```diff
@@ -0,0 +1,560 @@
+field. If we're lucky, that may lead them to intuit that the privacy is for the
+field, not the parameter.
+
+#### Provide a teaching error if they use `_` for other named parameters
+
+If a user tries to put `_` before a named parameter that *isn't* an initializing
+formal or declaring parameter, it's an error. That error message can explain
+that it's forbidden *here* but that the syntax can be used to declare or
+initialize a private field.
+
+#### Provide a teaching error if they use `_` on the argument name
+
+If a user sees a named parameter with a private name, they may try to call the
+constructor with that same private argument name, like `C(_variable: 123)`.
+When they do, this is always an error.
+
+The error message for that can explain that the `_` is only used to make the
+corresponding field private and that the argument should be the public name. The
+first time a user tries to call one of these constructors the wrong way, we can
+teach them the feature.
+
+### Super parameters
+
+We allow private named parameters for initializing formals and (assuming primary
+constructors exist), declaring parameters. What about the other special kind of
+constructor parameter, super parameters (the `super.` syntax)?
+
+Those are unaffected by this proposal. A super parameter generates an implicit
+argument that forwards to a superclass constructor. The super constructor's
+argument name is always public, even if the corresponding constructor parameter
+uses this feature and has a private name. Thus, super parameters continue to
+always use public names. For example:
+
+```dart
+class Tool {
+  int _price;
+
+  Tool({this._price}); // Private name here.
+}
+
+void cheapTool() => Tool(price: 1); // Called with public name.
+
+class Hammer extends Tool {
+  Hammer({super.price}); // And thus call with public name here too.
+}
+
+void pricyHammer() => Hammer(price: 200);
+```
+
+## Static semantics
+
+An identifier is a **private name** if it starts with an underscore (`_`),
+otherwise it's a **public name**.
+
+A private name may have a **corresponding public name**. If the characters of
+the identifier with the leading underscore removed form a valid identifier and a
+public name, then that is the private name's corresponding public name. *For
+example, the corresponding public name of `_foo` is `foo`.* If removing the
+underscore does not leave something which is is a valid identifier *(as in `_`
+or `_2x`)* or leaves another private name *(as in `__x`)*, then the private name
+has no corresponding public name.
+
+Given a named initializing formal or field parameter (for a primary constructor)
+with private name *p* in constructor C:
+
+*   If *p* has no corresponding public name *n*, then compile-time error. *You
+    can't use a private name for a named parameter unless there is a valid
+    public name that could be used at the call site.*
+
+*   If any other parameter in C has declared name *p* or *n*, then
```

Comment:
Should this say, "If any other _named_ parameter..."?

I don't see the conflict in `C(int p, {this._p});`. But we could still call it an error, if we want.

Reply:
It's deliberate but I agree it's subtle. Added some explanatory text.